### ISSUE TITLE
Architecture Analysis

### ISSUE THREAD

**Issue by harshsikka** [ManifoldRG/NEKO_Archive#2]
We need a concrete, objective analysis that presents the architectural detail in the GATO paper in a clearly understandable way.

**Comment by atharva-tendle** [comment]
Could you elaborate on what you mean by analysis @harshsikka? The way I see it we want to outline the different blocks on a high level for eg: embedding function, transformer decoder, masking, loss, etc., and then recursively hash out details for each of these blocks to the level of detail that someone can use it as a blueprint to convert into a working module.

Did you have something else in mind?

**Comment by harshsikka** [comment]
@atharva-tendle I think exactly this - we need someone to start with the conceptual/design overview for these blocks together and individually. I think the approach you outline is right!

TODO: break this out into the multiple necessary atomic design / planning tasks

**Comment by atharva-tendle** [comment]
We can sync more on Sunday, I have some thoughts regarding this.

**Comment by daniellawson9999** [comment]
In addition to architectural details of what we will build (gato specific components like tokenization), do we also want to think about what infrastructure we want to build on top of? What are the best open-source libraries for training large transformers? Is it something like https://github.com/EleutherAI/gpt-neox , research other options?

Another thing I have been thinking about is model initialization. While we probably want to investigate training from scratch, it may be worthwhile to consider finetuning an open-source LM.  I think this may work quite well. From the control task perspective, various works: [Can Wikipedia Help Offline Reinforcement Learning?](https://arxiv.org/abs/2201.12122) [On the Effect of Pre-training for Transformer in Different Modality on Offline Reinforcement Learning](https://openreview.net/forum?id=9GXoMs__ckJ), shows initializing with a LM (say GPT2) for learning control policies, is beneficial, such as decreases training time.
Maybe an example of what simple training infrastructure can look like for finetuning is [Stanford Alpaca](https://github.com/tatsu-lab/stanford_alpaca#fine-tuning), except we jointly finetune on language + all our other modalities. Their implementation seems relatively simple, based around huggingface transformers. This may be a way to start on some of our goals very efficiently.

**Comment by atharva-tendle** [comment]
Building on top of frameworks makes sense but honestly for a first pass simple pytorch + huggingface should be enough IMO. It would provide us with flexibility and the fact that the GATO setup seems quite straightforward I am wondering if we need to get too complex from the start. I am concerned about spending time debugging issues that repos built on top of torch and huggingface might come with. Have some more detailed thoughts on this that we can discuss during the sync.

Re: Training from scratch vs Finetuning.
- I think it's worth trying both for sure and something that we can scope into experiments. 
- Finetuning would be faster but it would come with its own sets of biases from the pre-training data and might have issues with adaptability.
- Training from scratch is slow and expensive but would mean that we are learning weights that are optimized keeping a multi-modal objective in mind.
- Both of these are essential!

**Comment by harshsikka** [comment]
Scoping this issue a bit more to reflect current project status:

• we have a somewhat clear plan on data gathering, control is a bit more clear but has a larger potential set of issues vs language + vision
• vision has another issue: do we include language grounded tasks that are amenable to seq2seq training objectives, or do we plan on including other objectives/datasets? 
• we have an idea of how tokenization and encoding works

We have a growing but fractured understanding of what the path forward is. Ideally, though it adds a bit of overhead, we want to have a clear experimental path and a more detailed research roadmap than the high level one [here](https://docs.google.com/document/d/1KE99nCOIymJtnF0nZ9D4G-wwQAGg7TsMG4hLSMEVBMs/edit?usp=sharing) (that we can of course tweak/iterate on)

I think the right move from a model & research design perspective, both to focus and move faster, is to do the following:

• Output 1 (in progress): review the whole paper, outline the architecture and each module in greater detail, and capture all questions/ambiguities/possibilities that we have raised in different issues, etc. Capture this in a doc for team review. 
• Output 2: following making decisions on the above, build a proper design doc for NEKO. capture iterative goals and experimental plan in/linked in this as well. to understand an ML system design document a bit better, check this out: 
• Output 3: granular experimental roadmap for this year. 

This issue's scope is really output 1, above. Will create issues for 2 (which will require several), and 3.

**Comment by harshsikka** [comment]
Closing issue for now, as it mostly served as a high level placeholder.

**Comment by henryj18** [comment]
About fine-tuning vs training from scratch, another aspect impacting the decision is the amount of compute available. Unless we can get large volume of compute, I would say whichever approach takes less computing resource should be the first choice, most likely the fine-tuning approach.

**Comment by harshsikka** [comment]
@henryj18 this is a good point, and something the team has discussed before briefly. The design space of potential solutions is significant, and other constraints also come into it (i.e. compute). Plus, our ability to access compute should grow as we continue to reach out. An experimental plan, given compute + training tricks + research aim is necessary.

**Comment by harshsikka** [comment]
@daniellawson9999 is thinking through training some small models on just the 2-3 control dataset combinations. Daniel might be a good idea to capture those ideas in an issue
